### DISCUSSION TITLE
Error: "could not find function 'readGDX'" when using gdx2 package in R

### DISCUSSION THREAD

**Discussion by Wutttong** [magpiemodel/magpie#722] [Help]
Dear MAgPIE Community,

I'm running into an issue with the readGDX function in the gdx2 package. When I try to use the function, I get the following error:
Error in readGDX(gdx, "f10_land", "f_land", format = "first_found") :
  could not find function "readGDX"

Here are the steps I’ve followed so far:
Installed the latest version of the gdx2 package from GitHub using devtools::install_github("pik-piam/gdx2").
Verified that gdx2 is loaded, and running ls("package:gdx2") shows that the readGDX function exists in the package.

However, when calling the function like this:
gdx2::readGDX(gdx, "f10_land", "f_land", format = "first_found")
I get the error: could not find function 'readGDX'.

In addition to this, I noticed that there may be some function masking happening due to other loaded packages (dplyr, magclass, etc.), but I am not sure if this is contributing to the issue. The console also shows the following:

The following object is masked from 'package:magclass':
    where
The following objects are masked from 'package:stats':
    filter, lag
The following objects are masked from 'package:base':
    intersect, setdiff, setequal, union

I suspect that the function is not being correctly invoked despite the package loading successfully. I'm not sure if this is a namespace conflict or an issue with the installation of readGDX.

I appreciate your time and look forward to your suggestions.

Thank you.

**Comment by tscheypidi** [comment]
hi @Wutttong,

can you check whether the package `gamstransfer` is properly installed? Normally it should be automatically installed when `gdx2` is being installed, but maybe something went wrong here. The reason I am asking is that gdx2::readGDX is internally using gamstransfer::readGDX and the error message might just indicate that it does not find the latter.

**Reply by Wutttong** [reply]
Hi Jan,

Yes, I have checked that gamstransfer is installed, but I am still encountering the issue with the readGDX function. It seems like gamstransfer is not being utilized correctly, or something else is causing the function to fail. Any further suggestions would be appreciated.

**Reply by tscheypidi** [reply]
can you create a minimal reproducible example to get this error? I could try to run it on my machine to see whether I receive the same error

**Reply by Wutttong** [reply]
Hi Jan,

I encountered an error while running the following code:

library("gdx2")
gdx <- "G:/0R3_NbS/MAgPIE/magpie/output/default_2024-09-20_08.57.38/fulldata.gdx"
result <- readGDX(gdx, "f10_land", "f_land", format = "first_found")
print(result)

I suspect that the issue arises from having both gdx and gdx2 packages installed, as they both contain the readGDX function, leading to a conflict. I attempted to remove the gdx package to resolve this conflict, but I still cannot successfully call readGDX from gdx2 in the script, despite being able to do so within RGui.

Could this issue be related to the fact that the model running script doesn't explicitly load library(gdx2)? I'd appreciate any advice on this.

Thanks in advance!

**Reply by tscheypidi** [reply]
I ran your example and it is working fine for me. But after I ran it I performed `install.packages("gdx2")` to make sure that I have the newest package installed. After that I immediatly run readGDX again and got the same error. After restarting the session it was working fine again. Could it be that you run `install.packages("gdx2")` immediatly before you read in the data? If not, I have currently no other idea what could cause the problem.

**Reply by pascal-sauer** [reply]
having both gdx and gdx2 installed at the same time shouldn't matter as long as not both are attached via `library`, but even then there should be a warning, and you should get either readGDX from gdx or from gdx2, but not "cannot find readGDX". In any case I'd recommend to call it like `gdx2::readGDX` and omit `library(gdx2)` to remove any ambiguity.

**Reply by pascal-sauer** [reply]
> the model running script doesn't explicitly load library(gdx2)

Which model running script do you mean here exactly? Did you also run the following in isolation (in a fresh R session)?
```
library("gdx2")
gdx <- "G:/0R3_NbS/MAgPIE/magpie/output/default_2024-09-20_08.57.38/fulldata.gdx"
result <- readGDX(gdx, "f10_land", "f_land", format = "first_found")
print(result)
```

**Reply by Wutttong** [reply]
> having both gdx and gdx2 installed at the same time shouldn't matter as long as not both are attached via `library`, but even then there should be a warning, and you should get either readGDX from gdx or from gdx2, but not "cannot find readGDX". In any case I'd recommend to call it like `gdx2::readGDX` and omit `library(gdx2)` to remove any ambiguity.

Thank you very much for your suggestion. I agree that using gdx2::readGDX could be an effective solution. However, I seem to be having trouble finding where to make this change in the code.

I appreciate your help and guidance on this matter.

**Reply by Wutttong** [reply]
> Which model running script do you mean here exactly? Did you also run the following in isolation (in a fresh R session)?

I attempted to include library(gdx2) in the start.R script, but it doesn't seem to have any effect. Additionally, I ran the example code in a fresh R session, and there were no errors.
